src/memcached/cache_impl.go

```diff
@@ -0,0 +1,300 @@
+// The memcached limiter uses GetMulti() to check keys in parallel and then does
+// increments asynchronously in the backend, since the memcache interface doesn't
+// support multi-increment and it seems worthwhile to minimize the number of
+// concurrent or sequential RPCs in the critical path.
+//
+// Another difference from redis is that memcache doesn't create a key implicitly by
+// incrementing a missing entry. Instead, when increment fails an explicit "add" needs
+// to be called. The process of increment becomes a bit of a dance since we try to
+// limit the number of RPCs. First we call increment, then add if the increment
+// failed, then increment again if the add failed (which could happen if there was
+// a race to call "add").
+
+package memcached
+
+import (
+	"context"
+	"math"
+	"math/rand"
+	"strconv"
+	"sync"
+
+	"github.com/coocood/freecache"
+	stats "github.com/lyft/gostats"
+
+	"github.com/bradfitz/gomemcache/memcache"
+
+	logger "github.com/sirupsen/logrus"
+
+	pb "github.com/envoyproxy/go-control-plane/envoy/service/ratelimit/v3"
+
+	"github.com/envoyproxy/ratelimit/src/assert"
+	"github.com/envoyproxy/ratelimit/src/config"
+	"github.com/envoyproxy/ratelimit/src/limiter"
+	"github.com/envoyproxy/ratelimit/src/settings"
+)
+
+type rateLimitCache interface {
+	// Same as in limiter.RateLimitCache
+	DoLimit(
+		ctx context.Context,
+		request *pb.RateLimitRequest,
+		limits []*config.RateLimit) []*pb.RateLimitResponse_DescriptorStatus
+
+	// Waits for any lingering goroutines that are incrementing memcache values.
+	// This is used for unit tests, since the memcache increments happen
+	// asynchronously in the background.
+	Flush()
```

Comment:
Would it be better to add the Flush() method to the main interface and just have it be a NOP for redis? I think that would be cleaner? Would this make it easier to share the tests between the two implementations?